Prevent converting bytestring ids to unicode ids when reordering on python2.

deiferni opened the pull request:
This PR fixes an issue where ids of reordered content are converted to unicode on python 2. We now convert input ids when necessary and make sure the ordering stored on a container contains only bytestring object ids.

When changing object ordering via PATCH request we used to incorrectly store ids of reordered resources as unicode instead of a bytestring (on python2). This lead to mixed types being stored in the ordering annotations and subsequently mixed types being returned when calling `objectIds` of a container.

The problem only exists with python2 as python3 uses unicode internally.

Things we should probably discuss in this PR:
- Is the suggested upgrade-path feasible?
- The current suggestion pushes responsibility to upgrade ordering to the client, see suggestion in upgrade-guide.

Fixes #827.

mister-roboto commented:
@deiferni thanks for creating this Pull Request and help improve Plone!

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass.

Whenever you feel that the pull request is ready to be tested, either start all jenkins jobs pull requests by yourself, or simply add a comment in this pull request stating:

```
@jenkins-plone-org please run jobs
```

With this simple comment all the jobs will be started automatically.

Happy hacking!

deiferni commented:
@jenkins-plone-org please run jobs

lukasgraf commented on src/plone/restapi/tests/mixin_ordering.py:
```diff
@@ -195,6 +195,38 @@ def test_reorder(self):
             self.folder.contentIds(),
         )
 
+    def test_ordering_preserves_bytestring_obj_id(self):
```
`bytestring` -> `native_string`
(See https://portingguide.readthedocs.io/en/latest/strings.html#the-native-string)

lukasgraf commented on src/plone/restapi/tests/mixin_ordering.py:
```diff
@@ -195,6 +195,38 @@ def test_reorder(self):
+        for id_ in self.folder.objectIds():
+            self.assertIsInstance(id_, str)
+
+    def test_ordering_preserves_bytestring_subset_ids(self):
```
bytestring -> native_string

lukasgraf commented on src/plone/restapi/tests/test_upgrade_ordering.py:
```diff
@@ -0,0 +1,46 @@
+                "Document", id="doc" + str(x), title="Test doc " + str(x)
+            )
+
+    def test_upgrade_ensure_child_ordering_object_ids_are_bytestrings(self):
```
bytestrings -> native_strings

lukasgraf commented on src/plone/restapi/upgrades/ordering.py:
```diff
@@ -0,0 +1,40 @@
+import six
+
+
+def ensure_child_ordering_object_ids_are_bytestrings(parent):
```
`bytestrings` -> `native_strings`
I would also prefer `container` instead of `parent` (parent of what?)

lukasgraf commented on docs/source/upgrade-guide.rst:
```diff
@@ -3,6 +3,24 @@ Upgrade Guide
+- See https://github.com/plone/plone.restapi/issues/827 for more details.
+
+We expect that most content won't actually be affected. Nevertheless in
+https://github.com/plone/plone.restapi/commit/6a3258d771bdb8dc91f693c181f6f09d14d680d6
```
-> `plone.restapi.upgrades.ordering`

deiferni commented:
@jenkins-plone-org please run jobs

tisto commented:
@deiferni what would be the possible downside of running the upgrade step by default?

deiferni commented:
@tisto it would be a potentially long-running upgrade for all content. We expect most content to be unaffected by the issue. Also we expect low impact for affected content. IMHO running the upgrade for all content by default poses a larger risk.

So as far as i can tell if we would want to be sure to fix all potentially affected content we would have to run the upgrade for every resource where theoretically content could have been reordered. That would most likely mean loading most/all content and perform the same duck-typing checks as in:https://github.com/plone/plone.restapi/blob/d00494f31a90a5d5070bea5e6f96254413148f54/src/plone/restapi/deserializer/mixins.py#L45-L52
Or maybe take a shortcut and load all `IFolderish` or something like that. But still for larger sites that would mean loading potentially thousands of objects to check if the ordering is broken (which it most likely is not as we expect most content has not been reordered via API).

Also as far as we can tell now the mixed types in the ordering annotations do not cause a default plone site to malfunction. IIRC we already discussed it a bit at plone conf and came to the conclusion that it is most likely not worth the risk to write and run an upgrade for all content. But if other users run into issues we can provide an upgrade helper which can be called with potentially affected content objects. And leave it to the clients to narrow down or define content which is potentially affected for them.

FYI we broke it in client code by relying on output of `objectIds` for traversal.

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
am a bit confused with the test errors. i think it is something unrelated to my changes, at least according to the failed tests.

deiferni commented:
@jenkins-plone-org please run jobs

tisto commented on docs/source/upgrade-guide.rst:
```diff
@@ -3,6 +3,23 @@ Upgrade Guide
 
 This upgrade guide lists all breaking changes in plone.restapi and explains the necessary steps that are needed to upgrade to the lastest version.
 
+
```
@deiferni we should add instructions here how to actually run the migration step (via ZMI). The alternative would be to provide a browser view that runs the upgrade step.

deiferni replied:
@tisto ok, will amend the following:
- implement an upgrade that checks everything that `is_folderish` according to the catalog (currently i only have a function that fixes one single object, the intention was that clients could implement their own upgrade. i see that we should provide a bit more convenience here) 
- browser view to manually run the upgrade

lukasgraf commented on docs/source/upgrade-guide.rst:
```diff
@@ -3,6 +3,23 @@ Upgrade Guide
+converts ids of reordered content to unicode from a bytestring when running on
+python2.
+
+- Only content that was reordered is affected.
```
Thinking of it, this could also do with some clarification. We should make it easy and unambigous for people to determine whether or not they need to run the fix or not. Maybe something like this:

```
You may be affected by this issue and should run the fix if:

- You used the PATCH "ordering" functionality of plone.restapi
- Were using Python 2 at that point
- Are seeing issues with objectIds() returning mixed string types
```

deiferni commented:
For the record, i have verified what happens if we do a `zodbupdate` from python2 to python3 with content with mixed string types in the ordering annotations. The `zodbupdate` script seems to be able to get rid of the mixed typing and correctly converts all the ids to native strings (i.e. `str` in python3), so upgrading from python2 to python3 will also get rid of potential issues. I think that reinforces the point not to run an upgrade by default.

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@tisto i have updated this PR with the following changes:
- added a browser-view in case someone needs the fix and prefers to run it TTW
- improved the description in the upgrade-guide, clarified when people are affected as suggested by @lukasgraf and also mentioned the two possible paths to run the fix

What is your stance on the upgrade strategy? We don't object rewriting this as a "proper" upgrade-step if someone does see a good reason to do so. Currently we think that not running the upgrade by default is the better strategy for the following reasons:
- we expect most content is not affected by the issue
- it only surfaces on python2
- even if content is affected it does not AFAICT have any serious consequences (unless you do very specific things with `contentIds` in client code)
- as soon as you migrated to python3 the issue will disappear, the current version of `zodbupdate` seems to update the object ids to native strings (i.e. unicode strings) correctly
- running the upgrade for all content by default may pose a certain risk if many objects need to be migrated, especially due to the potentially long runtime and loading of many objects into memory.

So from our side this would be good to 🚢. If anyone has objections please let me know so we can address/discuss the issue.

tisto commented:
@deiferni thank you! I will try to find time in the next days to review this and I will also discuss the upgrade issue in the FWT and with @esteele. My personal opinion is that the browser view is sufficient.

tisto commented:
@deiferni @lukasgraf sorry for not handling this in an appropriate timeframe. I promise to do a release early next week.

tisto commented:
@deiferni @lukasgraf plone.restapi 6.0.0 released with the fix. Sorry it took so long.

lukasgraf commented:
Thanks @tisto! :)
